design_documents/accessing_emitted_events.md

```diff
@@ -0,0 +1,191 @@
+# Accessing emitted events
+
+<!-- TOC -->
+* [Accessing emitted events](#accessing-emitted-events)
+  * [Context](#context)
+  * [Goal](#goal)
+  * [Considered Solutions](#considered-solutions)
+  * [`expect_events` cheatcode](#expectevents-cheatcode)
+    * [Usage example](#usage-example)
+  * [`spy_events` cheatcode](#spyevents-cheatcode)
+    * [Usage example](#usage-example-1)
+<!-- TOC -->
+
+## Context
+
+Some contract functions can emit events. It is important to test if they were emitted properly.
+
+## Goal
+
+Propose a solution that will allow checking if events were emitted.
+
+## Considered Solutions
+
+1. `expect_events` cheatcode 
+2. `spy_events` cheatcode
+
+## `expect_events` cheatcode
+
+Introduce a cheatcode with the signature:
+
+```cario
+fn expect_events(events: Array<snforge_std::Event>)
+```
+
+where `snforge_std::Event` is defined as below
+
+```cario
+struct Event {
+    name: felt252,
+    keys: Array<felt252>,
+    data: Array<felt252>
+}
+```
+
+- `name` is a name of an event passed as a shortstring,
+- `keys` are values under `#[key]`-marked fields of an event,
+- `data` are all other values in emitted event.
+
+`expect_event` cheatcode will define events which should be emitted in the next call. Other calls will not be affected.
+If provided events will not be emitted it will panic with a detailed message.
+
+`events` are the subsequence of all events emitted by the function, but you can also require function
```

Comment:
Subsequence suggests that they have to occur in that particular order, is that what you meant? If not, I would call it `subset`. Either way I would consider implementing both variants if that's possible (I guess we shall implement the `subset` variant now and leave `subsequence` for the future?)